engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/StraightThroughProcessingLoopValidator.java

```diff
@@ -0,0 +1,222 @@
+   *       element before. As a result we can conclude that we are in a loop.
+   *   <li>The element is not part of the potentialLoop list. We haven't passed the element before
+   *       so we are not in a loop yet. However, the element is a straight-through processing
+   *       element. We must check the outgoing sequence flow of this element to see if it loops back
+   *       around to an element that is part of the potentialLoop list.
+   *   <li>We have not encountered this element before, nor is it a straight-through processing
+   *       element. We can conclude that this path is not part of a straight-through processing
+   *       loop.
+   * </ul>
+   *
+   * @param potentialLoop a list which keeps track of elements that are part of a potential loop
+   * @param element the element used to determine if we reached a loop, or are definitely not part
+   *     of a loop
+   * @return an Either of which the left contains a list of the straight-through processing loop.
+   */
+  private static Either<List<ExecutableFlowNode>, ?> checkForStraightThroughProcessingLoop(
+      final LinkedList<ExecutableFlowNode> potentialLoop, final ExecutableFlowNode element) {
+
+    if (foundLoop(potentialLoop, element)) {
+      // It could happen that we detect a loop which doesn't involve the first element we checked.
+      // By taking a sublist we make sure we've isolated the loop.
+      final var loop = potentialLoop.subList(potentialLoop.indexOf(element), potentialLoop.size());
+      // This element is added to the loop twice. This is useful when creating an error message
+      // describing the loop.
+      loop.add(element);
+      return Either.left(loop);
+    } else if (STRAIGHT_THROUGH_PROCESSING_ELEMENT_TYPES.contains(element.getElementType())) {
+      // We are not in a loop yet, but the element is a straight-through processing element. We must
+      // keep checking for loops by analysing the outgoing sequence flows of this element.
+      potentialLoop.addLast(element);
+      Either<List<ExecutableFlowNode>, ?> isPartOfLoop = Either.right(null);
+      for (final ExecutableSequenceFlow outgoing : element.getOutgoing()) {
+        isPartOfLoop = checkForStraightThroughProcessingLoop(potentialLoop, outgoing.getTarget());
+        if (isPartOfLoop.isLeft()) {
+          break;
+        }
+      }
+
+      if (isPartOfLoop.isRight()) {
+        potentialLoop.remove(element);
+      }
+      return isPartOfLoop;
+
+    } else {
+      // No loops have been found. We can return an Either.Right to indicate success.
+      return Either.right(null);
+    }
+  }
+
+  /**
+   * This method will decide if we are in a loop or not. It does so by checking if the potentialLoop
+   * list contains the element we are currently checking. It also has an extra check to make sure we
+   * don't detect loops between just gateways. As of the time of writing it is unclear whether we
+   * should reject these or not.
```

Comment:
💭 I played around a bit and found that I can deploy a process that just consists of gateways which I was wondering whether we can reject that in general, not a big deal and no need to act imho.

Reply:
That's true, I explicitly made it an exception with the list of whitelisted elements types. Frankly, I wasn't sure if there were any viable use-cases and I didn't want to make it too strict. I figured those processes have been possible for a long time already without causing issues. Might as well keep allowing them until we see them start causing them 🤷